### ISSUE TITLE
NewCheck: UseEnhancedSwitch

### ISSUE THREAD

**Issue by mahfouz72** [checkstyle/checkstyle#15180]
child of https://github.com/checkstyle/checkstyle/issues/14961:

 We need to discuss the implementation of a new check that suggests using `->` (arrow syntax) instead of `:` (colon syntax) 

---

1. `->` (arrow syntax)
    - No need for break statements.
    - No fall-through behavior by design.
    - Directly returns a value without the need for braces and `yields` in switch expressions.
 
2. `:` (colon syntax) 
    - Requires break statements to prevent fall-through.
    - Fall-through behavior is possible if break statements are omitted.
    
---

Best Practices 

Using `->` :
- Preferable in switch expressions where you want to directly return a value.
- less error-prone since it avoids the need for break statements and eliminates the unintended fall-through behavior.

Using `:` : 
- Allows for fall-through behavior if that is desired.
- Suitable for cases where we want to group multiple case labels.

----
**Config**

```XML
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="UseEnhancedSwitch"/>
    </module>
</module>
```
---

**Examples**

```Java
public class Test {
  void t1(int x)  {
      switch (x) {           // violation, Switch can be replaced with enhanced switch
          case 1: {} break;
          case 2: {} break;
          case 3:{} break;
      }
      switch (x) {    // ok switch with fall through is ignored
          case 1: {} 
          case 2: {} break;
          case 3:{} break;
      }
      switch (x) {
          case 1 -> {}
          case 2 -> {}
          case 3 -> {}
      }
  }

  int t2(int x) {
      int y =  switch (x) {    // violation, Switch can be replaced with enhanced switch
          case 1 : yield 1;
          case 2 : yield 2;
          case 3 : yield 3;
          default: yield 0;
      };
      int yy = switch (x) {
          case 1 -> 1;
          case 2 -> 2;
          case 3 -> 3;
          default -> 0;
      };
      return y;
  }
}
```

---
**Output**
```
$ java -jar checkstyle-10.19.0-all.jar -c config.xml Test.java 
Starting audit...
[ERROR] D:\Test.java:3:8: Switch using reference types should have a null case. [UseEnhancedSwitch]
[ERROR] D:\Test.java:21:16: Switch using reference types should have a null case. [UseEnhancedSwitch]
Audit done.
Checkstyle ends with 2 errors.
```

**Comment by nrmancuso** [comment]
@mahfouz72 pleaes check with IDEA inspections, and see if they have a similar check. If so, let's explore the behavior of their check to gain some insights on how we can implement ours.

**Comment by mahfouz72** [comment]
https://www.jetbrains.com/help/inspectopedia/EnhancedSwitchMigration.html

I played with it a bit it is triggered on any switch that uses `:` and with no fall through

```Java
void t1(int o) {
        switch (o) {  // Switch statement can be replaced with enhanced 'switch'
            case 1:
                System.out.println("one");
                break;
            case 2:
                System.out.println("two");
                break;
            default:
                System.out.println("many");
        }
        switch (o) {  // not detected
            case 1:
                System.out.println("one");
            case 2:
                System.out.println("two");
                break;
            default:
                System.out.println("many");
        }
    }

    double getPrice(int c) {

        switch (c) { // Switch statement can be replaced with enhanced 'switch'
            case 0:
            case 1:
              return 0;
            case 2:
                System.out.println(2);
                return 2;
            case 3:
            case 4:
                 System.out.println(3);
                 return 3;
            case 5:
            default:
                 System.out.println(2);
        }
        
        switch (c) { // not detected
            case 0:
            case 1:
              return 0;
            case 2:
                System.out.println(2);
            case 3:
            case 4:
                 System.out.println(3);
                 return 3;
            case 5:
            default:
                 System.out.println(2);
        }
        return 0;
    }
```

**Comment by rnveach** [comment]
Looking at the bytecode, there is no difference between `:` and `->`.

**Comment by nrmancuso** [comment]
I am prepared to approve this one, @mahfouz72 please update the OP to show our checks behavior with proposed message.

**Comment by mahfouz72** [comment]
@nrmancuso done

**Comment by octylFractal** [comment]
I am working on this: https://github.com/checkstyle/checkstyle/pull/18489

**Comment by romani** [comment]
Check is merged.

**Comment by romani** [comment]
@octylFractal , you should have in 13.3.0, released today

**Comment by romani** [comment]
@octylFractal, are you interested to make auto fix for this Check ?

**Comment by octylFractal** [comment]
> you should have in 13.3.0, released today

🎉 

> are you interested to make auto fix for this Check ?

I don't think I have time for that right now, unfortunately.

**Comment by romani** [comment]
https://github.com/checkstyle/checkstyle-openrewrite-recipes/issues/154
